hpx/lcos/spmd_block.hpp

```diff
@@ -0,0 +1,275 @@
+
+                F()(hpx::launch::sync,
+                    hpx::find_here(),
+                    std::move(block),
+                    std::forward<Ts>(ts)...);
+            }
+        };
+
+        // Overload for lambdas
+        template <typename F>
+        struct spmd_block_helper<F,false>
+        {
+            std::string name_;
+            std::size_t num_images_;
+
+            template <typename ... Ts>
+            void operator()(std::size_t image_id, Ts && ... ts) const
+            {
+                using first_type =
+                    typename hpx::lcos::detail::extract_first_parameter<
+                                decltype(&F::operator())>::type;
+
+                static_assert(std::is_same<spmd_block, first_type>::value,
+                    "define_spmd_block() needs a lambda that " \
+                    "has at least a spmd_block as 1st argument");
+
+                spmd_block block(name_, num_images_, image_id);
+
+                int * dummy = nullptr;
+                hpx::util::invoke(
+                    reinterpret_cast<const F&>(*dummy),
+                    std::move(block),
+                    std::forward<Ts>(ts)...);
+            }
+        };
+    }
+
+    // Helper for define_spmd_block()
+    namespace detail
+    {
+        template <typename F, typename ... Args>
+        struct spmd_block_helper_action
+        {
+            static void call(
+                std::string name,
+                std::size_t images_per_locality,
+                std::size_t num_images,
+                Args... args)
+            {
+                using executor_type =
+                    hpx::parallel::execution::parallel_executor;
+
+                executor_type exec;
+                std::size_t offset = hpx::get_locality_id();
+                offset *= images_per_locality;
+
+                hpx::parallel::executor_traits<
+                    executor_type
+                    >::bulk_execute(
+                        exec,
+                        detail::spmd_block_helper<F>{name,num_images},
+                        boost::irange(
+                            offset, offset + images_per_locality),
+                        args...);
+            }
+        };
+    }
+
+    template <typename F, typename ... Args>
+    hpx::future<void>
+    define_spmd_block(std::string && name, std::size_t images_per_locality,
+        F && f, Args && ... args)
+    {
+        using ftype = typename std::decay<F>::type;
+
+        using helper_type =
+            hpx::lcos::detail::spmd_block_helper_action<
+               ftype, typename std::decay<Args>::type...>;
+
+        using helper_action_type =
+            typename hpx::actions::make_action<
+                decltype( &helper_type::call ), &helper_type::call >::type;
+
+        // Initialize f at compile-time in case when f is a lambda function
+        hpx::lcos::detail::initialize_if_lambda(f);
```

Comment:
This doesn't really work. It only might work if `f` has no state. This limits the usefulness of `spmd_block` tremenedously.The nullptr dereference and that you don't transfer the callable to the remote is just calling for trouble.